airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +47,162 @@ public void checkpoint() {
         LOG.debug("Database checkpoint complete.");
     }
 
-    @PreDestroy
-    public void onDestroy() {
-        Connection conn = null;
+    /**
+     * Return the current version of the HSQLDB database, as reported by the database properties file.
+     */
+    private String getHsqldbDatabaseVersion() {
+        Properties prop = new Properties();
+        File configFile = new File(settingsService.getDefaultJDBCPath() + ".properties");
+        if (!configFile.exists()) {
+            LOG.info("HSQLDB database doesn't exist and will be created on first run.");
+            return null;
+        }
+        try (InputStream stream = new FileInputStream(configFile)) {
+            prop.load(stream);
+            return prop.getProperty("version");
+        } catch (IOException e) {
+            LOG.error("Failed to determine HSQLDB database version.", e);
+            return null;
+        }
+    }
+
+    /**
+     * Return the paths to all existing database files.
+     */
+    private String[] getHsqldbDatabaseFiles() {
+        String[] extensions = { ".properties", ".script", ".log", ".data", ".backup", ".lck" };
+        return Arrays.stream(extensions)
+                .map(extension -> settingsService.getDefaultJDBCPath() + extension)
+                .filter(path -> new File(path).exists())
+                .toArray(size -> new String[size]);
+    }
+
+    /**
+     * Check if a HSQLDB database upgrade is needed.
+     * @return true if a database migration should be performed
+     */
+    private boolean isHsqldbDatabaseUpgradeNeeded() {
+        // Check the current database version
+        String currentVersion = getHsqldbDatabaseVersion();
+        if (currentVersion == null) {
+            return false;
+        }
+
+        // Check the database driver version
+        String driverVersion = null;
+        try {
+            Driver driver = DriverManager.getDriver(settingsService.getDefaultJDBCUrl());
+            driverVersion = String.format("%d.%d", driver.getMajorVersion(), driver.getMinorVersion());
+            if (driver.getMajorVersion() != 2) {
+                LOG.warn("HSQLDB database driver version {} is untested. Trying to connect anyway, this may upgrade the database from version {}.", driverVersion, currentVersion);
+                return false;
```

Comment:
I think we should set this true?
This setting determines whether a backup gets made.
So, in my opinion, any scenario where we're unsure of what's about to happen, we should make a backup, right?

Reply:
Done.